design-proposals/klip-11-DQL.md

```diff
@@ -1 +1,745 @@
+  * all inputs are tables
+  * at least one input is a stream
+* we don't consider the output type, because it's not a property of the query itself, but derived by the design of query language
+* we don't consider transient (client/CLI) queries vs. persistent queries (CSAS, CTAS) because both should have the same semantics
+
+Note, that while we consider those properties as independent in general, there are some corner case for which some properties contradict each other,
+i.e., there are some partial dependencies. More details below.
+
+Before we dive into a detailed discussion, we lay out (in less technical terms) what type of queries we want to support (this is for sure an incomplete list):
+
+* forever running query over an input STREAM that returns an output stream
+  * think: simple stream filter, stream-table join, or stream-stream join
+  * this may be a transient client query or a persistent query
+* forever running aggregation query over an input STREAM that returns an output table
+  * think: a persistent continuous query (may be windowed)
+  * for a persistent query, this should populate a MV
+  * for a transient client query, we would like to get the corresponding changelog stream
+* a RDBMS style query against a TABLE that terminates automatically and returns a table (called 'pull queries' in KLIP-8)
+  * conceptually, we can think of everything that 'classic' SQL allows
+* a self-terminating query over an input STREAM that returns an output stream
+  * think of a query with a range in time, e.g., yesterday, or everything up to "now" (but don't wait for future events)
+* a forever running query over an input TABLE that return a changelog stream
+  * answer my query on the current table state, and keep updating the answer if the base table state changes
+
+### Baseline Assumptions
+
+* Data stored in Kafka topics is just data; there is no semantic interpretation attached
+* A _STREAM_ is an unbounded append only-sequence of immutable facts
+* A _TABLE_ is a mutable bounded set
+  * for each point in (event) time, there exist exactly one TABLE _version_ (i.e., a snapshot)
+  * a TABLE _evolves_ over (event) time (we call it an _evolving table_) when new insert/updates/deletes are applied
+    * a TABLE version is "complete" if all updates for the corresponding event time are applied
+    * updating a TABLE from version v to v+1 might imply multiple updates and might take some time,
+      hence, at any point in wall-clock time a TABLE might be in an inconsistent state
+      (we call those the _generation_ of a table, i.e., for each table version, there might be multiple generation and only the last generation is consistent).
+  * a TABLE can be represented as (versioned) state or changelog – we consider the physical representation as an implementation detail
+
+
+
+## What is in scope
+
+The focus is to remove `EMIT CHANGES` and introduce new syntax to support different types of queries over streams and tables.
```

Comment:
> The focus is to remove `EMIT CHANGES` 

Humm... 

The point of adding the `EMIT` syntax was to control _materialization_ of the result. Initially we introduced `EMIT CHANGES`, but conceptually this contrasted `EMIT FINAL`, even if it wasn't in the syntax.  

[KLIP-10](https://github.com/confluentinc/ksql/pull/3754) is proposing to introduce `EMIT FINAL`.  How does this KLIP interplay with KLIP-10?

Is this how you see it working: 

|                        | EMIT syntax           | Proposed Syntax  |
| ------------- |--------------| -----|
| get value(s) from table | SELECT * FROM MY_TABLE  [EMIT FINAL] | SELECT * FROM MY_TABLE  |
| get stream of changes from a table | SELECT * FROM MY_TABLE  EMIT CHANGES | STREAM(SELECT * FROM MY_TABLE )
| windowed persistent query that outputs intermediate results per window      | SELECT COUNT() FROM MY_STREAM WINDOW WINDOW TUMBLING (SIZE 30 SECONDS) EMIT CHANGES      |   SELECT COUNT() FROM MY_STREAM WINDOW WINDOW TUMBLING (SIZE 30 SECONDS) |
| windowed persistent query that outputs only final result per window      | SELECT COUNT() FROM MY_STREAM WINDOW WINDOW TUMBLING (SIZE 30 SECONDS) EMIT FINAL      |   TABLE(SELECT COUNT() FROM MY_STREAM WINDOW WINDOW TUMBLING (SIZE 30 SECONDS)) |

How about when if/when we introduce the extensions to the `EMIT` syntax, as covered by the klip, e.g. rate limiting?

Reply:
That is an excellent point and my thoughts on it are not final. However, if you consider the `stream()` operator as different syntax to `emit changes` (not 100% if this holds up exactly), we could extend the `stream()` operator to embed all those "rate control" mechanism. Basically, both `stream()` and `emit changes` are some kind of CDC operation on the table.

There is also a difference between queries against streams and tables IMHO, and IMHO `emit changes` only applies to push/stream queries. The `emit changes` (and `emit final`) keyword does not really apply to pull/table queries IMHO and thus should not be supported/allowed for this case (it would also not be compliant with ANSI SQL that we aim to support from my understanding). 

I am also wondering, if `EMIT FINAL` is actually a property of the query, a property of the CDC operator, or the property of the window-aggregation-operator itself. To me, it seems that CDC rate control and `emit final` are two different things and that `final` is a property of the window-aggregation-operator itself. If this holds up, using `EMIT CHANGES` and `EMIT FINAL` syntax would convolute two concepts that should be separated. *I want to point out, that I am not 100% sure about this point---just wanted to share my current understanding.*

The first three examples for "proposed syntax" sounds about right. For the last one, the result table would still be updated with every update because the "inner query" return the full changelog stream. Also, the statement would only make sense as persistent `CTAS` query. A transient query would return an empty table as it returns the current table state when the query is issued and this table state would be empty when the query is issued. (compare section https://github.com/confluentinc/ksql/blob/888691db3d43e508ee1d87095288166117b501e2/design-proposals/klip-11-DQL.md#advanced-query-patterns that discussed this point in more details)

Hence, if the `final` property is a property of the window-aggregation-operator, it should not be part of the `stream()` operator (or `EMIT` clause). And for the last example, we would need to figure out how we change, eg, `WINDOW TUMBLING (SIZE 10 SECONDS)` syntax to allow specifying the `final` property.

Reply:
Copied from KLIP-10 comment for context https://github.com/confluentinc/ksql/pull/3754#issuecomment-559804952

> @mjsax
> 
> From the discussion in KLIP-8, the proposition was that for a given query, e.g.
> 
> ```sql
> SELECT * FROM FOO WHERE ROWKEY=x;
> ```
> 
> (Where FOO is a table thats computing an aggregate, e.g. `CREATE TABLE FOO AS SELECT COUNT() FROM BAR GROUP BY ROWKEY;`)
> 
> You can either request the changes of how the final result was built, using `EMIT CHANGES`, or only want the final result, using `EMIT FINAL`, with `EMIT FINAL` being the default if none explicitly supplied.
> 
> `EMIT CHANGES` would output the stream of CDC events of the history of how the count, for this key, changed over time. (Obviously, key compaction will _delete_ history, but that's configurable).
> 
> `EMIT FINAL` would get the current final result for the key. (Either from a state store, or potentially from reading the source topic and computing the aggregate. Using the state store is just an optimisation).
> 
> So in this sense both `EMIT CHANGES` and `EMIT FINAL` are valid.
> 
> Of course, that's not taking into account KLIP-11, which proposing using `SELECT * FROM STREAM(SELECT * FROM FOO WHERE ROWKEY=x);`.
> 
> However, this klip currently stands alone, or rather sits on top of klip-8. As such the `EMIT FINAL` is actually implicit for pull queries. Of course, we can decide if that should continue to be the case, but I think that is better discussed in klip-11.

Reply:
Escentially, what I think we're saying is that:
  *  `STREAM()` would be the same as `EMIT CHANGES`, (though current proposal is that `stream()` does not expose history CDC events, which we _may_ want), .
  * `TABLE()` would be the same as `EMIT FINAL`

> I am also wondering, if EMIT FINAL is actually a property of the query, a property of the CDC operator, or the property of the window-aggregation-operator itself. 

I don't think it's a feature of the window aggregation operator.  The window-aggregation-operator could be expressed as a table function, e.g.

```sql
SELECT * 
FROM hoppingWindow(
    myStream,
    myStream.eventTime,
    SIZE 30 SECONDS
)
```

which then outputs a table, (1 or more rows), per input event that reflected all the windows the input row falls into. (The schema of the output would be the schema of the stream, plus columns for window bounds).  However, to me, doing so demonstrates that the materialization control `EMIT` is providing is not part of the window-agg-op, but of the outer query itself, i.e. the query as an input that is the output of the table function, and the result materialization is performed on the query output, not the table function output.